### PR TITLE
Basic Node and ExpressJS API server with jest tests

### PR DESCRIPTION
**Pull request by aknutas**
Basic Node and ExpressJS API server structure with GET support and jest+supertest tests.

Resolve issue #20 
Resolve issue #21

### CODE REVIEW

**Comment by aknutas** [comment]
It's not perfect and has some cruft here and there, but the basics are there.

File: .gitignore [hunk 1]
```diff
@@ -48,4 +48,21 @@ node_modules/
 .next
 
 # MACOS sillyness
-.DS_Store
\ No newline at end of file
+.DS_Store
+
+# WebStorm project files
```
Comment (lukekearney) [inline .gitignore:53]: I think we should be able to just ignore all .idea directories

File: src/APIServer/app.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
+const createError = require('http-errors');
+const express = require('express');
+const path = require('path');
+const cookieParser = require('cookie-parser');
+const logger = require('morgan');
+const Promise = require('bluebird'); // eslint-disable-line no-unused-vars
+const mongoose = require('mongoose');
+
+mongoose.Promise = Promise;
+// const viewEngine = require('express-json-views');
```
Comment (lukekearney) [inline src/APIServer/app.js]: If it's not needed then remove

File: src/APIServer/app.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+// Load app
+const app = express();
+
+// Loading routers
+const noticesRouter = require('./routes/notices');
+
+// Set view engine, JSON (not in use yet)
+// See https://www.npmjs.com/package/express-json-views
+// app.engine('json', viewEngine());
```
Comment (lukekearney) [inline src/APIServer/app.js]: Are these documentation or old code? If the latter then please remove

File: src/APIServer/app.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// app.set('view engine', 'json');
+
+// DB connection
+const db = require('./model/db'); // eslint-disable-line no-unused-vars
+const notice = require('./model/notice'); // eslint-disable-line no-unused-vars
+
+// view engine setup
+// No views required yet
+// app.set('views', path.join(__dirname, 'views'));
+// app.set('view engine', 'pug');
```
Comment (lukekearney) [inline src/APIServer/app.js]: Same as above

File: src/APIServer/app.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+// error handler
+app.use((err, req, res, next) => { // eslint-disable-line no-unused-vars
+  // set locals, only providing error in development
+  res.locals.message = err.message;
+  res.locals.error = req.app.get('env') === 'development' ? err : {};
+
+  // render the error page
+  res.status(err.status || 500);
+  // res.render('error');
```
Comment (lukekearney) [inline src/APIServer/app.js]: As above

File: src/APIServer/bin/www [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ */
+
+server.listen(port);
+server.on('error', onError);
+server.on('listening', onListening);
+
+/**
+ * Normalize a port into a number, string, or false.
+ */
+
```
Comment (lukekearney) [inline src/APIServer/bin/www]: Here and elsewhere, remove the line separating docs comments from the function

File: src/APIServer/bin/www [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * Event listener for HTTP server "error" event.
+ */
+
+function onError(error) {
+  if (error.syscall !== 'listen') {
+    throw error;
+  }
+
+  var bind = typeof port === 'string'
```
Comment (lukekearney) [inline src/APIServer/bin/www:55]: I'm not sure how I feel about ternary operators here. We'll leave them for now

File: src/APIServer/model/db.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+const mongoose = require('mongoose');
+
+mongoose.connect('mongodb://127.0.0.1/TWbackend').then(() => {
```
Comment (lukekearney) [inline src/APIServer/model/db.js:3]: Is there any way we can use the config file values rather than a hardcoded string?

File: src/APIServer/bin/www [hunk 9]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env node
```
Comment (lukekearney) [inline src/APIServer/bin/www:1]: Is this file necessary, or is it automatically generated? if it is automatically generated, should we be committing it?

**Review by lukekearney — CHANGES_REQUESTED** [review]
Just some basic comments for now. No idea how this pull request got so big.

After this change merges it might be worth talking about reduced pull request sizes...but I need to think about that a bit

File: .gitignore (hunk 1 above)
Comment (aknutas) [reply .gitignore:53]: Edited; replaced the lines with .idea/ (this should be the syntax according to documentation, feel free to correct...)

File: src/APIServer/app.js (hunk 2 above)
Comment (aknutas) [reply src/APIServer/app.js]: Revised across the board

File: src/APIServer/bin/www (hunk 6 above)
Comment (aknutas) [reply src/APIServer/bin/www]: Done

File: src/APIServer/bin/www (hunk 9 above)
Comment (aknutas) [reply src/APIServer/bin/www:1]: It has been generated as part of the default ExpressJS template and is the "default execution script". I think it is worth leaving in. We do need some standard execution path, port setting and others.

File: src/APIServer/model/db.js (hunk 8 above)
Comment (aknutas) [reply src/APIServer/model/db.js:3]: I will look into it in the next commit and combine it with the bin/www somehow. The hardcoding is a cludge, but I'd prefer to get this out and then looking into merging the /bin/www and config file or command line configs.

**Comment by aknutas** [comment]
There we go, revised version committed; based on review comments.

File: src/APIServer/app.js (hunk 3 above)
Comment (aknutas) [reply src/APIServer/app.js]: Revised

File: src/APIServer/app.js (hunk 5 above)
Comment (aknutas) [reply src/APIServer/app.js]: Revised

File: src/APIServer/bin/www (hunk 9 above)
Comment (lukekearney) [reply src/APIServer/bin/www:1]: Oh ok, so it is autogenerated on run? It's probably pulling from the config anyway then (see my earlier comment).

Comment (aknutas) [reply src/APIServer/bin/www:1]: No. In the earlier comment you refer to there is a false dichotomy: Automatically generated or useful. It can be automatically generated and useful. It has been generated by the ExpressJS project template generator when the project was created and is supposed to be the "executable" for the project. If it is deleted we are not getting it back and I cannot easily rewrite its contents from scratch.

Comment (lukekearney) [reply src/APIServer/bin/www:1]: Ah I see. Thanks

**Comment by lukekearney** [comment]
Antti, how do I actually run this? At the moment I'm getting a Mongo network error. I'm sure I could figure it out but it's probably easier just to ask you.

Is it possible an npm script that we can run like `npm run api` that will start everything required by the APIserver?

Maybe it's for a follow up pull request, but could we also document how to run it? Do I need Mongo installed and running locally etc

**Comment by aknutas** [comment]
Yes. Documentation is definitely a priority in addition to a better startup script. I think documentation is already an issue in GitHub and we can add a startup script as another issue as well. I eventually aspire to have a Docker script or image as well so that people can easily get a prepackaged environment.

You need to have a MongoDB server running for the tests. On OS X you can do "brew install mongo" and then I believe for example "mongod --dbpath ~/tmp/mongotwdatabase/".

**Comment by aknutas** [comment]
Addendum: Starting up Mongo automatically is not possible because it is platform dependent. For example in Linux it is automatically running and in OS X and Windows the startup commands vary. This can be solved with a Docker container, though, which I will set up next after this merge is approved. It will come with a prepackaged environment.

Edit: Docker Hub has autobuild based on GitHub, so that is perfect in that regard.

File: .gitignore (hunk 1 above)
Comment (lukekearney) [reply .gitignore:53]: I think it should be ok now

File: src/APIServer/bin/www (hunk 7 above)
Comment (lukekearney) [reply src/APIServer/bin/www:55]: This is a generated file and so ternary operators or not doesn't matter. Ignoring

File: src/APIServer/model/db.js (hunk 8 above)
Comment (lukekearney) [reply src/APIServer/model/db.js:3]: No worries. If the file is autogenerated then we could be OK. Think we can leave it for now

**Comment by lukekearney** [comment]
After reading Antti's replies I think we should be OK to merge. I seem to have broken the build by trying to resolve merge conflicts though :(

**Review by lukekearney — APPROVED** [review]
